### PR TITLE
Add version info

### PR DESCRIPTION
**Pull request by teamwong111**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Related issue: https://github.com/open-mmlab/mmcv/issues/976
Add GCC version, Cuda version, torch version and git hash in version.py. Honestly, I am not quite sure about stability on Linux, Windows and macOS.

## Modification

Impl related functions. 

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

When run `pip install [option] .` or `python setup.py [option]`, some version info will be added in `version.py`. It may look like this
```python
GCC = '7.5.0'
CUDA = '11.2.78'
PyTorch = '1.5.1'
CommitID = 'ad71e62'
```

According to our Github CI, we can see it is more easy to get these info.
```shell
Run python -c "import mmcv; print(f'GCC {mmcv.version.gcc_version}', f'CUDA {mmcv.version.cuda_version}', f'PyTorch {mmcv.version.torch_version}', f'CommitID {mmcv.version.commit_id}', sep='\n')"
GCC 7.5.0
CUDA None
PyTorch 1.5.1
CommitID ad71e62
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
We can print the version information in CI.

- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L46
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L80
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L119
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L186
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L244
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L276
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L305
- https://github.com/open-mmlab/mmcv/blob/add157cc73b05c7be3d9b63505e17dc2288a6d0f/.github/workflows/build.yml#L349

File: setup.py [hunk 1]
```diff
@@ -19,6 +24,187 @@
...
+
+        return sha
+
+    mmcv_root = Path(__file__).parent
+    version_path = mmcv_root / 'mmcv' / 'version.py'
+    gcc_version = get_gcc()
+    cuda_version = get_cuda()
+    git_hash = get_git_hash()
+    if EXT_TYPE:
+        torch_version = torch.__version__
```
Comment (zhouzaida) [inline setup.py:193]: The value `torch.__version__` is different between `pip install torch` and `conda install torch`. More details at https://github.com/open-mmlab/mim/blob/62bd756d7045c3c591d442415c72a8308e778207/mim/utils/utils.py#L321

File: setup.py [hunk 2]
```diff
@@ -19,6 +24,187 @@
...
+    else:
+        torch_version = None
+    with open(version_path, 'a') as f:
+        f.write(f'\n'
+                f"gcc_version = '{gcc_version}'\n"
+                f"cuda_version = '{cuda_version}'\n"
+                f"torch_version = '{torch_version}'\n"
+                f"git_version = '{git_hash}'\n")
```
Comment (zhouzaida) [inline setup.py:201]:
```suggestion
                f"commit_id = '{git_hash}'\n")
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -346,7 +356,7 @@ jobs:
           rm -rf .eggs
           CC=clang CXX=clang++ CFLAGS='-stdlib=libc++' pip install -e .
       - name: Validate the installation
-        run: python -c "import mmcv"
+        run: python -c "import mmcv; print(f'GCC {mmcv.version.gcc_version}', f'CUDA {mmcv.version.cuda_version}', f'PyTorch {mmcv.version.torch_version}', f'Git {mmcv.version.commit_id}', sep='\n')"
```
Comment (zhouzaida) [inline .github/workflows/build.yml:359]: Git -> CommitID

File: setup.py [hunk 4]
```diff
@@ -19,6 +24,189 @@
...
+        except OSError:
+            sha = fallback
+
+        return sha
+
+    mmcv_root = Path(__file__).parent
+    version_path = mmcv_root / 'mmcv' / 'version.py'
+    gcc_version = get_gcc()
+    cuda_version = get_cuda()
+    git_hash = get_git_hash()
```
Comment (zhouzaida) [inline setup.py:191]:
```suggestion
    git_hash = get_git_hash(digits=7)
```
We just need to keep the first 7 characters.

File: setup.py [hunk 5]
```diff
@@ -19,6 +24,189 @@
...
+            sha = out.strip().decode('ascii')
+            if digits is not None:
+                sha = sha[:digits]
+        except OSError:
+            sha = fallback
+
+        return sha
+
+    mmcv_root = Path(__file__).parent
+    version_path = mmcv_root / 'mmcv' / 'version.py'
```
Comment (zhouzaida) [inline setup.py:188]:
```python
import os.path as osp
...
version_path = osp.join(mmcv_root, 'mmcv', 'version.py')
```

And replace all the `os.path` with `osp`.

File: setup.py [hunk 6]
```diff
@@ -19,6 +24,189 @@
...
+            for name in fnames:
+                if os.access(name,
+                             os.F_OK | os.X_OK) and not os.path.isdir(name):
+                    return name
+        return None
+
+    def find_nvcc():
+        nvcc = which('nvcc')
+        if nvcc is not None:
+            return os.path.dirname(nvcc)
```
Comment (zhouzaida) [inline setup.py:62]:
```python
import os.path as osp
```

Replace all the `os.path` with `osp`.

File: setup.py [hunk 7]
```diff
@@ -19,6 +24,189 @@
     print('Skip building ext ops due to the absence of torch.')
 
 
+def add_version_info():
+    """Modified from
+    https://github.com/pytorch/pytorch/blob/v1.3.1/tools/setup_helpers."""
```
Comment (zhouzaida) [inline setup.py:29]:
```suggestion
    # Modified from
    # https://github.com/pytorch/pytorch/blob/v1.3.1/tools/setup_helpers.
```

File: setup.py [hunk 8]
```diff
@@ -19,6 +24,189 @@
...
+                         default).upper() in ['ON', '1', 'YES', 'TRUE', 'Y']
+
+    def check_negative_env_flag(name, default=''):
+        return os.getenv(name,
+                         default).upper() in ['OFF', '0', 'NO', 'FALSE', 'N']
+
+    def which(thefile):
+        path = os.environ.get('PATH', os.defpath).split(os.pathsep)
+        for d in path:
+            fname = os.path.join(d, thefile)
```
Comment (zhouzaida) [inline setup.py:48]:
```suggestion
            fname = osp.join(d, thefile)
```

File: setup.py [hunk 9]
```diff
@@ -19,6 +24,189 @@
...
+        if check_negative_env_flag('USE_CUDA') or check_env_flag('USE_ROCM'):
+            cuda_version = None
+        else:
+            if IS_LINUX or IS_DARWIN:
+                CUDA_HOME = os.getenv('CUDA_HOME', LINUX_HOME)
+            else:
+                CUDA_HOME = os.getenv('CUDA_PATH', '').replace('\\', '/')
+                if CUDA_HOME == '' and len(WINDOWS_HOME) > 0:
+                    CUDA_HOME = WINDOWS_HOME[0].replace('\\', '/')
+            if not os.path.exists(CUDA_HOME):
```
Comment (zhouzaida) [inline setup.py:118]:
```suggestion
            if not osp.exists(CUDA_HOME):
```

File: setup.py [hunk 10]
```diff
@@ -19,6 +24,189 @@
...
+                # We use nvcc path on Linux and cudart path on macOS
+                if IS_LINUX or IS_WINDOWS:
+                    cuda_path = find_nvcc()
+                else:
+                    cudart_path = ctypes.util.find_library('cudart')
+                    if cudart_path is not None:
+                        cuda_path = os.path.dirname(cudart_path)
```
Comment (zhouzaida) [inline setup.py:125]:
```suggestion
                        cuda_path = osp.dirname(cudart_path)
```

File: setup.py [hunk 11]
```diff
@@ -19,6 +24,189 @@
...
+                    else:
+                        cuda_path = None
+                if cuda_path is not None:
+                    CUDA_HOME = os.path.dirname(cuda_path)
```
Comment (zhouzaida) [inline setup.py:129]:
```suggestion
                    CUDA_HOME = osp.dirname(cuda_path)
```

File: setup.py [hunk 12]
```diff
@@ -19,6 +24,189 @@
...
+        torch_version = None
+    with open(version_path, 'a') as f:
+        f.write(f'\n'
+                f"gcc_version = '{gcc_version}'\n"
+                f"cuda_version = '{cuda_version}'\n"
+                f"torch_version = '{torch_version}'\n"
+                f"commit_id = '{git_hash}'\n")
+
+
+if 'egg_info' not in sys.argv:
```
Comment (zhouzaida) [inline setup.py:206]: Please add a comment to describe which command will make this condition true?

File: .github/workflows/build.yml (hunk 3 above)
Comment (teamwong111) [reply .github/workflows/build.yml]: done

File: setup.py (hunk 4 above)
Comment (teamwong111) [reply setup.py]: done

File: setup.py (hunk 6 above)
Comment (teamwong111) [reply setup.py]: done

File: setup.py (hunk 7 above)
Comment (teamwong111) [reply setup.py]: done

File: setup.py (hunk 12 above)
Comment (teamwong111) [reply setup.py:209]: done

**Review by zhouzaida — APPROVED** [review]

File: setup.py [hunk 13]
```diff
@@ -19,6 +25,191 @@
...
+
+    mmcv_root = Path(__file__).parent
+    version_path = mmcv_root / 'mmcv' / 'version.py'
+    gcc_version = get_gcc()
+    cuda_version = get_cuda()
+    git_hash = get_git_hash(digits=7)
+    if EXT_TYPE:
+        torch_version = torch.__version__
+        if '+' in torch_version:  # eg. 1.8.1+cu111 -> 1.8.1
+            torch_version = torch_version.split('+')[0]
```
Comment (zhouzaida) [inline setup.py:195]: Maybe the torch_version should be the format like `1.8.1+cu111`.

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -43,7 +43,7 @@ jobs:
       - name: Build and install
         run: rm -rf .eggs && pip install -e .
       - name: Validate the installation
-        run: python -c "import mmcv"
+        run: python -c "import mmcv; print(f'GCC {mmcv.version.gcc_version}', f'CUDA {mmcv.version.cuda_version}', f'PyTorch {mmcv.version.torch_version}', f'CommitID {mmcv.version.commit_id}', sep='\n')"
```
Comment (ZwwWayne) [inline .github/workflows/build.yml:46]: we can directly call collect_env.py for simplicity

**Comment by ZwwWayne** [comment]
May post the results in the PR message, e.g., what is the new verion.py looks like? Furthermore, can we do it more easily by using the existing collect_env?

File: .github/workflows/build.yml (hunk 14 above)
Comment (teamwong111) [reply .github/workflows/build.yml:46]: We run this command just to check whether the version info we get is correct on different OS.

**Comment by teamwong111** [comment]
> May post the results in the PR message, e.g., what is the new verion.py looks like? Furthermore, can we do it more easily by using the existing collect_env?

The results have been posted in the PR message. We cannot use the existing collect_env because collect_env relies on C++ or CUDA which will work after build.

**Comment by zhouzaida** [comment]
@ZwwWayne 

We need to record the compilation information in `version.py`, because it doesn't make sense to record it in `mmcv.ops`. Users may encounter `undefine symbol` errors when importing operators from `mmcv.ops`, and this error is usually due to inconsistencies between the cuda version used to compile mmcv-full and the runtime version, so we need the user to provide the compile-time cuda information, which is not available from mmcv.ops due to the `undefine symbol` errors. So it is necessary to record the cuda information in `version.py`.

So did you mean that "Can we simplify the way to get the gcc and cuda versions using `from mmcv.ops import get_compiler_version, get_compiling_cuda_version`"

File: setup.py (hunk 13 above)
Comment (teamwong111) [reply setup.py]: done

**Comment by ZwwWayne** [comment]
> 

yes

**Comment by zhouzaida** [comment]
**Windows**
```
Traceback (most recent call last):
  File "setup.py", line 220, in <module>
    add_version_info()
  File "setup.py", line 189, in add_version_info
    gcc_version = get_gcc()
  File "setup.py", line 138, in get_gcc
    stderr=subprocess.PIPE)
  File "D:\miniconda3\envs\py37-pt18-cu111\lib\subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "D:\miniconda3\envs\py37-pt18-cu111\lib\subprocess.py", line 1207, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] 系统找不到指定的文件。
```

File: setup.py [hunk 15]
```diff
@@ -20,6 +25,201 @@
...
+                torch_version += '+cu'
+                torch_version += (torch.version.cuda).replace('.', '')
+            else:
+                torch_version += '+cpu'
+    else:
+        torch_version = None
+    with open(version_path, 'r+') as f:
+        lines = f.readlines()
+        last_line = lines[-1]
+        if '__all__' not in last_line:
```
Comment (zhouzaida) [inline setup.py:205]: It maybe better to put those information after `__version__` rather than `__all__`

File: setup.py [hunk 16]
```diff
@@ -20,6 +25,201 @@
...
+                    CUDA_HOME = osp.dirname(cuda_path)
+                else:
+                    CUDA_HOME = None
+            cuda_version = find_cuda_version(CUDA_HOME)
+        return cuda_version
+
+    def get_gcc():
+        proc = subprocess.Popen(['gcc', '--version'],
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
```
Comment (zhouzaida) [inline setup.py:138]: image

**Pull request closed without merging**